atc/worker/gardenruntime/image.go

```diff
@@ -40,19 +38,48 @@ func (worker *Worker) fetchImageForContainer(
 	delegate runtime.BuildStepDelegate,
 ) (FetchedImage, error) {
 	logger := lagerctx.FromContext(ctx)
+
 	if imageSpec.ImageArtifact != nil {
-		volume, ok, err := worker.findVolumeForArtifact(ctx, teamID, imageSpec.ImageArtifact)
+		volume, err := worker.findOrStreamVolume(ctx, imageSpec.Privileged, teamID, container, imageSpec.ImageArtifact, "/", delegate)
 		if err != nil {
-			logger.Error("failed-to-locate-artifact-volume", err)
+			logger.Error("failed-to-find-or-stream-volume-for-image", err)
 			return FetchedImage{}, err
 		}
-		if ok && volume.DBVolume().WorkerName() == worker.Name() {
-			// it's on the same worker, so it will be a baggageclaim volume
-			volumeOnWorker := volume.(Volume)
-			return worker.imageProvidedByPreviousStepOnSameWorker(ctx, logger, imageSpec.Privileged, teamID, container, volumeOnWorker)
-		} else {
-			return worker.imageProvidedByPreviousStepOnDifferentWorker(ctx, imageSpec.Privileged, teamID, container, imageSpec.ImageArtifact, delegate)
+
+		imageVolume, err := worker.findOrCreateCOWVolumeForContainer(
+			ctx,
+			imageSpec.Privileged,
+			container,
+			volume,
+			teamID,
+			"/",
+		)
+		if err != nil {
+			logger.Error("failed-to-create-cow-volume-for-image", err)
+			return FetchedImage{}, err
```

Comment:
nit! Just something that might be useful that we had in the `imageProvidedByPreviousStepOnSameWorker` method previously. If you think itll be helpful, can also add extra error printing info for error handling of `loadMetadata` and `StreamFile`.

```suggestion
			return FetchedImage{}, fmt.Errorf("create COW volume: %w", err)
```

Reply:
@evanchaoli I am not sure if @andy-paine could still pick this up, would you be able to update this PR to move it forward? thx

Reply:
@xtremerui I'm willing to move forward this PR, but I'm not aware of how to update other's PR. Only way I'm realizing is to check-pick all commits and create a new PR, but that sounds stealing @andy-paine 's work. The other way is that, as the comments are all nit issues, you just go ahead merge this PR, then I can use a new PR to address the comments. What's your suggestion?

Reply:
I think you can checkout this PR.

```
git checkout -b depop-avoid-duplicating-parallel-volume-streams master
git pull https://github.com/depop/concourse.git avoid-duplicating-parallel-volume-streams
```
Then you can create a new branch from it and make changes. Submit your branch as a new PR. All commit history will be there and we just need to close this PR and continue on yours.

Reply:
@xtremerui Okey, your way is better than cherry-pick. Following your way, I have created #8322.

@clarafu Addressed this comment in #8322.